File: agent/unity_tools/connection.py
"""
Shared connection utilities for Unity WebSocket communication.

This module provides the async bridge between LangGraph tools and the Unity WebSocket connection.
Tools call `call_unity_async()` which routes through the UnityManager's WebSocket.
"""
import json
import logging
from typing import Any, Optional

logger = logging.getLogger("movesia.unity_tools")

# Global reference to unity_manager - set during server startup
_unity_manager = None


def set_unity_manager(manager) -> None:
    """
    Set the global unity manager reference.
    Called from server.py during startup.
    """
    global _unity_manager
    _unity_manager = manager
    logger.info("Unity manager registered with tools")


def get_unity_manager():
    """Get the global unity manager instance."""
    return _unity_manager


async def call_unity_async(action: str, **params) -> dict:
    """
    Send a request to Unity via WebSocket and wait for response.

    Args:
        action: The Unity command type (e.g., 'get_hierarchy', 'create_gameobject')
        **params: Command parameters

    Returns:
        Response dict from Unity

    Raises:
        RuntimeError: If no Unity connection is available
    """
    if _unity_manager is None:
        raise RuntimeError("Unity manager not initialized. Tools cannot communicate with Unity.")

    if not _unity_manager.is_connected:
        return {
            "success": False,
            "error": "Unity is not connected. Please ensure Unity Editor is running and connected.",
            "hint": "Check that the Movesia plugin is installed in Unity and the WebSocket connection is established."
        }

    logger.debug(f"call_unity_async: action={action}, params={json.dumps(params)}")

    try:
        result = await _unity_manager.send_and_wait(action, **params)
        logger.debug(f"call_unity_async result: {str(result)[:200]}")
        return result
    except TimeoutError as e:
        logger.warning(f"Unity command timed out: {action}")
        return {
            "success": False,
            "error": f"Command timed out: {action}",
            "hint": "Unity may be busy (compiling, showing a dialog, etc.). Try again."
        }
    except RuntimeError as e:
        logger.error(f"Unity command failed: {action} - {e}")
        return {
            "success": False,
            "error": str(e)
        }
    except Exception as e:
        logger.error(f"Unexpected error in call_unity_async: {e}", exc_info=True)
        return {
            "success": False,
            "error": f"Unexpected error: {str(e)}"
        }

File: src/agent/unity-tools/connection.ts
/**
 * Shared connection utilities for Unity WebSocket communication.
 *
 * This module provides the async bridge between LangGraph tools and the Unity WebSocket connection.
 * Tools call `callUnityAsync()` which routes through the UnityManager's WebSocket.
 */

import type { UnityManager, UnityResponse } from './types';

/** Global reference to unity_manager - set during server startup */
let unityManager: UnityManager | null = null;

/**
 * Set the global unity manager reference.
 * Called from server.ts during startup.
 */
export function setUnityManager(manager: UnityManager): void {
    unityManager = manager;
}

/**
 * Get the global unity manager instance.
 */
export function getUnityManager(): UnityManager | null {
    return unityManager;
}

/**
 * Send a request to Unity via WebSocket and wait for response.
 *
 * @param action - The Unity command type (e.g., 'get_hierarchy', 'create_gameobject')
 * @param params - Command parameters
 * @returns Response from Unity
 * @throws Error if Unity manager not initialized
 */
export async function callUnityAsync(
    action: string,
    params: Record<string, unknown> = {}
): Promise<UnityResponse> {
    console.log(`[Unity Tool] Calling action: ${action}`, params);

    if (unityManager === null) {
        console.error('[Unity Tool] Unity manager not initialized');
        throw new Error('Unity manager not initialized. Tools cannot communicate with Unity.');
    }

    if (!unityManager.isConnected) {
        console.warn('[Unity Tool] Unity not connected - returning error response');
        return {
            success: false,
            error: 'Unity is not connected. Please ensure Unity Editor is running and connected.',
            hint: 'Check that the Movesia plugin is installed in Unity and the WebSocket connection is established.'
        };
    }

    try {
        console.log(`[Unity Tool] Sending command to Unity: ${action}`);
        const result = await unityManager.sendAndWait(action, params);
        console.log(`[Unity Tool] Received response for: ${action}`, { success: (result as any)?.success });
        // The UnityManager returns the message body which should have success/error fields
        // Cast to UnityResponse - Unity's plugin should always include these fields
        return result as unknown as UnityResponse;
    } catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);

        if (errorMessage.toLowerCase().includes('timeout')) {
            console.warn(`[Unity Tool] Command timed out: ${action}`);
            return {
                success: false,
                error: `Command timed out: ${action}`,
                hint: 'Unity may be busy (compiling, showing a dialog, etc.). Try again.'
            };
        }

        console.error(`[Unity Tool] Command failed: ${action} - ${errorMessage}`);
        return {
            success: false,
            error: errorMessage
        };
    }
}
